docs/README.md

```diff
@@ -1,20 +1,201 @@
+* `register_command`: takes as arguments the name of a command and a function which should execute when the command is received. The function is user defined, and takes an instance of `DAQModule::data_t` as argument. `DAQModule::data_t` is aliased to the `nlohmann::json` type and can thus be thought of as a blob of JSON-structured data. While in principle any arbitary name could be associated with any function of arbitrary behavior to create a command, in practice implementors of DAQ modules define commands associated with the DAQ's state machine: "_conf_", "_start_", "_stop_", "_scrap_". Not all DAQ modules necessarily need to perform an action for each of those transitions; e.g., a module may only be designed to do something during configuration, and not change as the DAQ enters the running state ("_start_") or exits it ("_stop_").  
+* `init`: this pure virtual function's implementation is meant to create objects which are persistent for the lifetime of the DAQ module. It takes as an argument the type `DAQModule::data_t`. Typically it will use parameters from this JSON argument to define the persistent objects. For persistent objects of types which don't have an efficient copy assigment operator, a common idiom to declare as member data a `unique_ptr` to the type of interest and then, in `init`, to allocate the desired object on the heap using values from the JSON and point the `unique_ptr` member to it. Queues are commonly allocated in `init`; they'll be described in more detail later in this document. 
+
+An conceptual example of what this looks like is the following simplified version of a DAQ module implementation. 
+```
+// This file would be called plugins/MyDaqModule.hpp
+// Functions would typically be defined in plugins/MyDaqModule.cpp
+
+class MyDaqModule : public dunedaq::appfwk::DAQModule {
+  public:
+     MyDaqModule(const std::string& name) : // A DAQ module instance is meant to have a unique name
+        dunedaq::appfwk::DAQModule(name)
+          {
+          register_command("conf",  &MyDAQModule::do_conf);
+          register_command("start", &MyDAQModule::do_start);
+          register_command("stop",  &MyDAQModule::do_stop);
+          register_command("scrap", &MyDAQModule::do_scrap);
+     }
+     
+     void init(const data_t& init_data) override;
+  
+  private:
+  
+     void do_conf(const data_t& conf_data);
+     void do_start(const data_t& start_data);
+     void do_stop(const data_t& stop_data);
+     void do_scrap(const data_t& scrap_data);
+};
+```
+A set of programming idioms developed over the first year of DAQ module development which, while not strictly necessary for implementing DAQ modules, have proven to be quite useful. They'll be described within the context of the functions above. Of course, you can also see them in action by looking at the source code of actual DAQ modules. 
```

Comment:
I think this sentence is talking about ThreadHelper, but it doesn't seem to belong here